app/code/community/HE/TwoFactorAuth/controllers/Adminhtml/TwofactorController.php

```diff
@@ -207,13 +219,23 @@ private function _verifyGoogle()
                         );
                     }
 
-                    $msg = Mage::helper('he_twofactorauth')->__("Valid code entered");
-                    Mage::getSingleton('adminhtml/session')->addSuccess($msg);
                     Mage::getSingleton('admin/session')->set2faState(HE_TwoFactorAuth_Model_Validate::TFA_STATE_ACTIVE);
                     $user = Mage::getSingleton('admin/session')->getUser();
                     $password = $user->getPassword();
-                    Mage::dispatchEvent('twofactor_auth_verification_success', array('password' => $password, 'user' => $user, 'result' => true));
-                    $this->_redirect('*');
+                    $lockInfo = new Varien_Object();
```

Comment:
This is a brilliant hack. 

Typically any values that are returned from a dispatched observer do not reach the controlller

We are setting a varien_object to record the locked state for an user, and using that to update the messaging on the frontend

More details here: https://stackoverflow.com/questions/13277121/return-data-back-to-dispatcher-from-event-observer-in-magento